Create configuration conditional "bench"

tcharding opened the pull request:
Currently we are unable to build with all features enabled with a non-nightly toolchain, this is because of the use of

    `#![cfg_attr(all(test, feature = "unstable"), feature(test))]`

which causes the following error when building:

 error[E0554]: `#![feature]` may not be used on the stable release channel

The "unstable" feature is used to guard bench mark modules, this is widely suggested online but there is a better way.

When running the bench marks use the following incantation:

    `RUSTFLAGS='--cfg=bench' cargo bench`

This creates a configuration conditional "bench" that can be used to guard the bench mark modules.

```
#[cfg(bench)]
mod benches {
    ...
}
```

Kixunil commented:
Perhaps `#[cfg(bench)] extern crate test;`?

apoelstra commented:
Very nice! Would be nice to also mention this in the README -- and we should move all the crates in the ecosystem over to this scheme.

apoelstra reviewed (approved):
ACK 56050115770b51f299258335fbd253566dff82cd

Kixunil commented on contrib/test.sh:
```diff
@@ -69,10 +69,10 @@ then
 if [ "$DO_BENCH" = true ]
 then
-    cargo bench --features unstable
+    RUSTFLAGS='--cfg=bench' cargo bench
```
Maybe `cargo +nightly bench`?

apoelstra replied:
This would make it impossible to use specific toolchains for benching, no?

Kixunil replied:
If you have multiple nightlies, yes (is that even possible?) Also I think `+beta` is also possible. Stable can't work because `extern crate test` is experimental.

apoelstra replied:
You can definitely have multiple nightlies; you specify them by date. This is sometimes useful for bisecting when a nightly broke things.

Understood that `stable` won't work...we should, I suppose, check whether the user has specified a toolchain and provide a helpful error message if he hasn't.

tcharding replied:
By specify you mean by using `TOOLCHAIN=foo`, right? We don't do that in `test.sh` so we can't warn for not being set.

apoelstra replied:
That is what I mean, and I don't understand what you mean by "we don't do that". We have a check for it on line 6.

tcharding replied:
Oh I mean we never set the `TOOLCHAIN` env variable in CI. We could check if `TOOLCHAIN` is set, and check what toolchain is being used, then output an error message if `TOOLCHAIN` is set and current toolchain is not nightly.

tcharding commented:
Changes in force push:
- Made `test.sh` comment be: `# Bench if told to, only works with non-stable toolchain (nightly, beta).`

tcharding commented:
> Perhaps #[cfg(bench)] extern crate test;?

Forgot to say the other day, you were right @Kixunil, thanks man!

tcharding commented:
Woops, readme suggestion still to come ...

tcharding commented:
Now includes readme section (and an initial patch fixing up a trivial markdown issue in the readme file).

tcharding commented:
I thought that cfg_docs was _not_ an experimental feature but I was wrong: https://doc.rust-lang.org/unstable-book/language-features/doc-cfg.html

Force push puts the comment about experimental features back to how it was and groups `cfg_docs` together with `test` as it was.

apoelstra reviewed (approved):
ACK 64d50b9a7a187f56a1713c36effd10ea7c45385c

Kixunil reviewed (approved):
ACK f3b2120ec92636759c08c9dfe105c6bb08766007

BTW if we renamed `TOOLCHAIN` to `RUSTUP_TOOLCHAIN` we could get rid of alias and possibly it'd be more obvious what it does/how it works.

tcharding commented:
> ACK [f3b2120](https://github.com/rust-bitcoin/rust-bitcoin/commit/f3b2120ec92636759c08c9dfe105c6bb08766007)
> 
> BTW if we renamed `TOOLCHAIN` to `RUSTUP_TOOLCHAIN` we could get rid of alias and possibly it'd be more obvious what it does/how it works.

Cool! Added to my todo list as I work on improving CI.

apoelstra reviewed (approved):
ACK f3b2120ec92636759c08c9dfe105c6bb08766007

apoelstra commented:
@tcharding anything more you wanna add to this PR, or should we merge it?

tcharding commented:
Its  good to go mate.
